### ISSUE TITLE
Make when policy is evaluated more easily configurable.

### ISSUE THREAD

**Issue by undergroundwebdesigns** [trailblazer/trailblazer#79]
Currently the Policy module over-rides the #setup! method to evaluate the policy as soon as other setup is complete. The issue that I'm running into is that I would like to use my policy to confirm that a user can only create Widgets that he owns, or where the widget's user_id matches the logged in user's id. With the current Policy implementation, since the policy is run immediately on setup, no params data has been synced to the model yet and so I can't do this check in the policy, which only knows about the model and the current user.

One idea I had to solve this would be to delay calling #evaluate_policy until after validations have passed, but _before_ the block is run, passing the validated contract to the policy instead of the model.

Another option would be to simply to treat it like a callback and require people to manually call some sort of authorize! method manually, at the place that suits them best. I like the flexibility of that later option, but don't like that it makes it all too easy to forget to run the policy check altogether.

**Comment by fernandes** [comment]
hi @undergroundwebdesigns , some points:

1. could open a PR with the test case you're trying to cover (or a gist), I tried to figure out but being honestly would be much more cleaner to see some code

2. As you told, as a callback is really easy to forget to call I personally don't like this approach for policy stuff

3. Suggestion: Policy it overrides [setup](https://github.com/apotonick/trailblazer/blob/ca7037658de8f96969dffb3128f2a4f82845c810/lib/trailblazer/operation/policy.rb#L32) and call super before evaluating the policy

And on Operation [setup](https://github.com/apotonick/trailblazer/blob/ca7037658de8f96969dffb3128f2a4f82845c810/lib/trailblazer/operation.rb) if run `setup_params!` and and `build_model!`

you can try to write a `setup_model!` on your Operation to be run before the policy, maybe a validate and contract.sync (its like the process, but would run before, once you need to setup you model before the policy)

**Comment by apotonick** [comment]
What I do here is:

1. I add a passthrough policy on the class level.
2. Somewhere in the depths of `#process` I then use the `policy` object to run my custom authorization logic.

```ruby
class Create < Trailblazer::Operation
  policy Comment::Policy, :pass # Policy has a method #pass returning true

  def process(params)
    # do whatever
    raise NotAuthorizedError if policy.admin? # this could be abstracted to Op#authorize!
  end
```

We could even allow to "disable" the check in `setup!` so you don't need this `pass` trick.

**Comment by fernandes** [comment]
great solution

I suggested the `setup_model!` because seems he needs to setup the model before the policy be evaluated, was more semantic to me heheh

**Comment by timoschilling** [comment]
The policy should be evaluated after `setup_model!`

**Comment by apotonick** [comment]
This is exactly what happens, @timoschilling: https://github.com/apotonick/trailblazer/blob/master/lib/trailblazer/operation/policy.rb#L34

@undergroundwebdesigns Have you tried the `pass` trick? Should we make that a default policy?

**Comment by undergroundwebdesigns** [comment]
The pass trick works, but seems a little awkward. At that point I might as well not use the Policy module at all, and just instantiate and call the policy class directly, or even mix in the Pundit library and use that interface.

`setup_params!`, `setup_model!` etc. don't work for what I'm trying to do. I'm using representers, and in that case there's not really any way to access the incoming parameters from the document until after the document is parsed in `validate`. 

Ideally, if there's going to be an automatic policy check when Policy is included (and I think there should be) I'd like to see the policy be instantiated with the contract instance, instead of the model, and be evaluated after validate, but before the validate block is run or validate returns. That gives the policy access to all the parsed attributes of the request, before anything is actually sync'd to the model.

**Comment by apotonick** [comment]
We can also allow to include `Policy::Builder` or something that won't change `setup!`.

Wait, using representers still allows using `setup_model!` since the id is usually encoded in the URL, so you grab the model, instantiate the contract and _then_ use the representer to deserialize the JSON document to the contract. Chapter 11 talks about that in detail, and I'm finishing some improvements in the involved gems as I type.

How do you use representers right now, @undergroundwebdesigns ?

**Comment by undergroundwebdesigns** [comment]
Yes, using representers allows using setup_model! in that the model will be created / looked up by id. However, the _new data_ will not be available in the model at that point, and thus won't be available to the policy.

As an example, say I'm writing an API endpoint to create widgets, and one of the parameters being sent is "user_id" or the id of the user that should own the widget being created. Using the current Policy code, my Widget::Create operation will instantiate a new (empty) widget model, pass that to the policy, and then run the policy `#create?` method. If my policy contains something like this:

```
def create?
   return user.present? && user.id == model.user_id
end
```
The policy will always fail, because at this point the model.user_id is always nil. Even if we were editing an existing widget, user_id might be set, so the policy might pass, but then the incoming data _may_ change the value, and the policy won't have any chance of catching this.

tl;dr: as currently coded, the policy has no way of taking into account pending changes to the model in question, just it's initial, un-changed state.

**Comment by fernandes** [comment]
you can setup the model on `setup_model!` and evaluate the policy on it, save on a variable and overwrite the `model!` method to get from `setup_model!` variable... make sense?

I'm just trying to suggest things to make your use case works then we thing on a proper solution on trailblazer

**Comment by apotonick** [comment]
@undergroundwebdesigns The class policy is meant to do exactly that: Make an initial authorization check. If you don't need that, you don't use it, but you can still benefit from the `policy` object throughout your operation (a bit like I use it in the book to decide whether or not to display fields in the form).

All that's missing is:

1. Allow to have a policy but don't run it in `setup!`.
2. Introduce an `authorize!(:create?)` method for the operation so it actually throws an exception.

```ruby
class Comment::Create < Trb::Op
  include Policy::Build # not sure if we need this at all.
  policy Comment::Policy # note how I don't specify an initial rule!

  def process(params)
    # do whatever you want
    authorize!(:create?) # raises NotAuthorizedError if invalid
  end
end
``` 

Is that what you need, @undergroundwebdesigns ?

**Comment by undergroundwebdesigns** [comment]
Yes, that would cover most of it. Can we code the authorize method to optionally take the "model" to authorize as a parameter? Something like `authorize!(:create?, model)`. The model argument could be optional, but I'd like to be able to authorize against the populated contract instead of the model, at least some of the time.

**Comment by apotonick** [comment]
That makes sense, but then we can't store the policy object and have to recreate it every time you use `authorize!` to avoid stale policy objects.

**Comment by apotonick** [comment]
Can you join our gitter channel, I have some important questions for you about the way you use policies? :grin:

**Comment by monkbroc** [comment]
I ended up creating a manual policy module where I need to explicitly call `authorize!` at the right moment.

```ruby
module Trailblazer
  # Adds #authorize! and ::policy.
  class Operation
    module ManualPolicy
      def self.included(includer)
        includer.extend DSL
      end

      module DSL
        def self.extended(extender)
          extender.inheritable_attr :policy_class
        end

        def policy(policy_class)
          self.policy_class = policy_class
        end
      end

      def policy
        @policy ||= self.class.policy_class.new(current_user, model)
      end

      private

      def authorize!(action)
        fail policy_exception(policy, action, model) unless policy.public_send(action)
      end

      def policy_exception(policy, action, model)
        NotAuthorizedError.new(query: action, record: model, policy: policy)
      end
    end
  end
end
```

I use it like this:

```ruby
  class Attachments::Create < Trailblazer::Operation
    include ManualPolicy
    policy AttachmentPolicy

    def process(params)
      validate(params[:attachment]) do |f|
        f.sync
        authorize! :create?
        model.save
      end
    end

    # other stuff like contract, setup_model!
  end
```

**Comment by apotonick** [comment]
In 2.0, you can run it whenever you want: http://trailblazer.to/gems/operation/2.0/policy.html#pundit

**Comment by monkbroc** [comment]
Keep up the good work @apotonick! Thanks!
